src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -94,13 +84,122 @@ module Writer = struct
         let%bind _ = Pipe.downstream_flushed t.root_pipe in
         Deferred.unit )
 
-  let close t =
+  let rec close : type a. a t -> unit =
+   fun t ->
     guard_already_closed t (fun () ->
         Pipe.close t.root_pipe ;
-        Int.Table.iter t.pipes ~f:(fun w -> Pipe.close w) ;
+        Int.Table.iter t.pipes ~f:(function
+          | Leaf writer -> Pipe.close writer
```

Comment:
I'm not sure if closing an already close pipe is safe, can you check? If it's not you need to call `Pipe.is_closed` or whatever here before trying to close.